### PR TITLE
12.0 last successful transaction

### PR DESCRIPTION
**Pull request by kga-odoo**

### CODE REVIEW

File: caldoor_payment/models/payment_acquirer.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        for token in self:
+            last_date = None
+            for payment in token.payment_ids:
+                if payment.state == 'done' and (not last_date or payment.date > last_date):
+                    last_date = payment.date
+            token.last_successful_transaction_date = last_date
+
+    @api.multi
+    def name_get(self):
+        # include_last_transaction_date = self._context.get('include_last_transaction_date', False)
```
Comment (kga-odoo) [inline caldoor_payment/models/payment_acquirer.py:22]: remove unused code

File: caldoor_payment/models/payment_acquirer.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        # if not include_last_transaction_date:
+        #     return super(PaymentToken, self).name_get()
```
Comment (kga-odoo) [inline caldoor_payment/models/payment_acquirer.py:24]: Make sure name_get for same model is not overriden anywhere else. In both odoo and enterprise addons.

File: caldoor_payment/models/payment_acquirer.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        names = []
+        for token in self:
+            token_name = token.name
+            if token.last_successful_transaction_date:
+                token_name += ' %s' % token.last_successful_transaction_date
```
Comment (kga-odoo) [inline caldoor_payment/models/payment_acquirer.py:26]: replace this with                
`token_name += ' %s %s' % (token.partner_id.name, token.last_successful_transaction_date)`

File: caldoor_payment/models/payment_acquirer.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+from dateutil import tz
+
+from odoo import api, fields, models
+
+
+class PaymentToken(models.Model):
+    _inherit = 'payment.token'
+
+    last_successful_transaction_date = fields.Datetime(compute='_compute_last_successful_transaction_date')
+    contact_name = fields.Char(string='Contact Name')
```
Comment (kga-odoo) [inline caldoor_payment/models/payment_acquirer.py:12]: Check in the staging given by customer .
Its is a related field. No need to make it editable.

Comment (kak-odoo) [reply caldoor_payment/models/payment_acquirer.py:12]: Yeah, I specifically asked, it's not supposed to be a related field, they just want a text field

File: caldoor_payment/models/payment_acquirer.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+            if token.last_successful_transaction_date:
+
+                token_name += ' %s' % token._get_formatted_time()
+            names.append((token.id, token_name))
+        return names
+
+    def _get_formatted_time(self):
+        self.ensure_one()
+        timezone = tz.gettz('America/Los_Angeles')
+        transaction_date_pacific = self.last_successful_transaction_date.astimezone(timezone)
```
Comment (kga-odoo) [inline caldoor_payment/models/payment_acquirer.py:39]: no need for manual timezone conversion.
https://github.com/odoo/odoo/blob/d66652b4aa514546356306e3eb48eaecaba07c33/odoo/fields.py#L1866
Use this instead.

**Pull request merged**